--- src/JDefs.h.old	2011-12-02 17:12:51.000000000 +1000
+++ src/JDefs.h	2016-11-19 15:59:33.406991513 +1000
@@ -33,7 +33,7 @@
 #endif
 
 /*
- * Default definitions (for GCC/Linux)
+ * Default definitions (for GCC/Linux)
  */
 #if defined _FILE_OFFSET_BITS && _FILE_OFFSET_BITS == 64
 #define JDIFF_LARGEFILE
@@ -76,7 +76,7 @@
 #define JDIFF_COPYRIGHT "Copyright (C) 2002-2005,2009,2011 Joris Heirbaut"
 
 #define uchar unsigned char
-#define ulong unsigned long int         // unsigned long
+typedef unsigned long int ULONG ;
 #define null  NULL
 
 #ifdef _LARGESAMPLE
